File: solutions/94.binary-tree-inorder-traversal/iteration.ts
/*
 * @lc app=leetcode.cn id=94 lang=typescript
 *
 * https://leetcode-cn.com/problems/binary-tree-inorder-traversal
 */

/**
 * Definition for a binary tree node.
 */
class TreeNode {
    val: number;
    left: TreeNode | null;
    right: TreeNode | null;
    constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
        this.val = val === undefined ? 0 : val;
        this.left = left === undefined ? null : left;
        this.right = right === undefined ? null : right;
    }
}

// @lc code=start
function inorderTraversal(root: TreeNode | null): number[] {
    const stack: TreeNode[] = [];
    const seqences: number[] = [];
    let current: TreeNode | null = root;
    while (current || stack.length > 0) {
        while (current) {
            stack.push(current);
            current = current.left;
        }
        current = stack.pop() as TreeNode;
        seqences.push(current.val);
        current = current.right;
    }
    return seqences;
}
// @lc code=end

export default inorderTraversal;

File: solution/13.roman-to-integer/iteration.rs
/*
 * @lc app=leetcode.cn id=13 lang=rust
 *
 * https://leetcode-cn.com/problems/roman-to-integer
 */

// @lc code=start
impl Solution {
    pub fn roman_to_int(s: String) -> i32 {
        let mut result = 0;
        let mut last = 0;
        for c in s.chars() {
            let current = match c {
                'I' => 1,
                'V' => 5,
                'X' => 10,
                'L' => 50,
                'C' => 100,
                'D' => 500,
                'M' => 1000,
                _ => 0,
            };
            if current > last {
                result += current - 2 * last;
            } else {
                result += current;
            }
            last = current;
        }
        result
    }
}
// @lc code=end
